### ISSUE TITLE
why 'acc' is getting worse after stacking? 

### ISSUE THREAD

**Issue by zhaokunyao** [vecxoz/vecstack#3]
task:       [classification]
n_classes:  [3]
metric:     [accuracy_score]
mode:       [oof_pred_bag]
n_models:   [4]

model 0:    [ExtraTreesClassifier]
    fold 0: [0.84220100]
    fold 1: [0.83712466]
    fold 2: [0.85259327]
    fold 3: [0.83750569]
    fold 4: [0.83925319]
    ----
    MEAN:   [0.84173556] + [0.00571701]
    FULL:   [0.84173644]

model 1:    [RandomForestClassifier]
    fold 0: [0.87585266]
    fold 1: [0.87170155]
    fold 2: [0.88307552]
    fold 3: [0.87437415]
    fold 4: [0.86930783]
    ----
    MEAN:   [0.87486234] + [0.00468014]
    FULL:   [0.87486349]

model 2:    [XGBClassifier]
    fold 0: [0.90995907]
    fold 1: [0.91446770]
    fold 2: [0.91128298]
    fold 3: [0.91169777]
    fold 4: [0.90983607]
    ----
    MEAN:   [0.91144872] + [0.00167472]
    FULL:   [0.91144885]

model 3:    [GradientBoostingClassifier]
    fold 0: [0.90950432]
    fold 1: [0.91037307]
    fold 2: [0.91765241]
    fold 3: [0.90350478]
    fold 4: [0.90391621]
    ----
    MEAN:   [0.90899016] + [0.00515841]
    FULL:   [0.90899163]

Final prediction score: [0.90554191]

**Comment by vecxoz** [comment]
There may be several reasons:

**[1] Hyperparameters for 2nd level model are not optimal.**

Did you perform cross-validation to choose hyperparameters? You can do it with the same tool. Given that you already have `S_train` and `S_test` (1st level) you can repeatedly run the following code with different hyperparameters (e.g. `n_estimators=100`, then `n_estimators=200`, etc.) and chose the best `MEAN` score. `S_test_level_2` would be your final 2nd level prediction corresponding to the best `MEAN` score.

```python
models_level_2 = [        
    XGBClassifier(random_state=0, n_jobs=-1, learning_rate=0.1, 
                  n_estimators=100, max_depth=3)
]

S_train_level_2, S_test_level_2 = stacking(models_level_2,
                                           S_train, y_train, S_test,
                                           regression=False,
                                           mode='oof_pred',
                                           random_state=0,
                                           verbose=2)
```

**[2] Sometimes majority voting (for classification with class labels) or weighted average (for regression or classification with probabilities) applied to 1st level predictions is better then some complicated 2nd level model like XGBoost.**

Given that you already have `S_train` and `S_test` (1st level) you can try to compute final prediction using majority voting:

```python
from scipy.stats import mode
y_pred = mode(S_test, axis=1)[0]
print(accuracy_score(y_test, y_pred))
```

**[3] 1st level models are highly correlated.**

Main stacking rule says that less correlated 1st level models give better result.
It means that you can always try some other 1st level models (even less accurate).
